install-manager/src/main/java/cm/aptoide/pt/install_manager/RealTask.kt

```diff
@@ -69,23 +77,47 @@ internal class RealTask internal constructor(
     return this
   }
 
+  @OptIn(ExperimentalCoroutinesApi::class)
   internal suspend fun start() {
     if (isFinished) return // Already finished.
-    val result = when (type) {
+    job = when (type) {
       Task.Type.UNINSTALL -> packageInstaller.uninstall(packageName)
         .onStart { _stateAndProgress.emit(Task.State.UNINSTALLING to -1) }
-        .collectErrorFor(Task.State.UNINSTALLING)
-
-      else -> checkSizeToState(sizeEstimator.getDownloadSize(installPackageInfo))
-        ?: packageDownloader.download(packageName, installPackageInfo)
-          .onStart { _stateAndProgress.emit(Task.State.DOWNLOADING to -1) }
-          .collectErrorFor(Task.State.DOWNLOADING)
-        ?: _stateAndProgress.emit(Task.State.READY_TO_INSTALL to -1).let { null }
-        ?: checkSizeToState(sizeEstimator.getInstallSize(installPackageInfo))
-        ?: packageInstaller.install(packageName, installPackageInfo)
-          .collectErrorFor(Task.State.INSTALLING)
-    } ?: Task.State.COMPLETED
-    _stateAndProgress.emit(result to -1)
+        .onEach { _stateAndProgress.emit(state to it) }
+
+      Task.Type.INSTALL -> {
+        listOf(
+          packageDownloader.download(packageName, installPackageInfo)
+            .onStart {
+              _stateAndProgress.emit(Task.State.DOWNLOADING to -1)
+              checkSizeToState(sizeEstimator.getDownloadSize(installPackageInfo))
+            }
+            .onEach { _stateAndProgress.emit(Task.State.DOWNLOADING to it) },
+          packageInstaller.install(packageName, installPackageInfo)
+            .onStart {
+              _stateAndProgress.emit(Task.State.READY_TO_INSTALL to -1)
+              checkSizeToState(sizeEstimator.getInstallSize(installPackageInfo))
+            }
+            .onEach { _stateAndProgress.emit(Task.State.INSTALLING to it) }
+        )
+          .asFlow()
+          .flattenConcat()
+      }
+    }
+      .onCompletion {
+        val result = when (it) {
+          null -> Task.State.COMPLETED
+          is AbortException -> Task.State.ABORTED
+          is CancellationException -> Task.State.CANCELED
+          is OutOfSpaceException -> Task.State.OUT_OF_SPACE
+          else -> Task.State.FAILED
+        }
+        _stateAndProgress.emit(result to -1)
+      }
+      .catch { /* Suppress failures */ }
```

Comment:
It's not necessary, because all the errors are mapped inside `onComlpete`.
This is needed just to not let flow crash the `start` method.
But if one needs to debug he can add this temporary.